File: utils/shuffle/shuffle_test.go
package shuffle_test

import (
	"strconv"
	"testing"

	"github.com/navidrome/navidrome/utils/shuffle"
	. "github.com/onsi/ginkgo/v2"
	. "github.com/onsi/gomega"
)

func TestShuffle(t *testing.T) {
	RegisterFailHandler(Fail)
	RunSpecs(t, "Shuffle Suite")
}

type track struct {
	id, artist, album string
}

func keys(t track) shuffle.Keys {
	return shuffle.Keys{Artist: t.artist, Album: t.album}
}

func ids(tracks []track) []string {
	out := make([]string, len(tracks))
	for i, t := range tracks {
		out[i] = t.id
	}
	return out
}

func clone(tracks []track) []track {
	out := make([]track, len(tracks))
	copy(out, tracks)
	return out
}

func adjacentSame(tracks []track, field func(track) string) int {
	n := 0
	for i := 1; i < len(tracks); i++ {
		if field(tracks[i]) != "" && field(tracks[i]) == field(tracks[i-1]) {
			n++
		}
	}
	return n
}

func minAdjacent(n, maxFreq int) int {
	if n == 0 {
		return 0
	}
	return max(0, 2*maxFreq-n-1)
}

var _ = Describe("IsRandomSort", func() {
	It("matches Native API, SQL random(), and SEEDEDRAND rewrite", func() {
		Expect(shuffle.IsRandomSort("random")).To(BeTrue())
		Expect(shuffle.IsRandomSort("RANDOM")).To(BeTrue())
		Expect(shuffle.IsRandomSort("random()")).To(BeTrue())
		Expect(shuffle.IsRandomSort("SEEDEDRAND('media_file|abc', media_file.id)")).To(BeTrue())
		Expect(shuffle.IsRandomSort("title")).To(BeFalse())
		Expect(shuffle.IsRandomSort("")).To(BeFalse())
		Expect(shuffle.IsRandomSort("order_title")).To(BeFalse())
	})
})

var _ = Describe("TrackKeys", func() {
	It("falls back to album artist when track artist is empty", func() {
		Expect(shuffle.TrackKeys("", "AA", "X")).To(Equal(shuffle.Keys{Artist: "AA", Album: "X"}))
		Expect(shuffle.TrackKeys("A", "AA", "X")).To(Equal(shuffle.Keys{Artist: "A", Album: "X"}))
	})
})

var _ = Describe("Slice", func() {
	It("leaves empty and single-item slices unchanged", func() {
		empty := []track{}
		shuffle.Slice(empty, keys)
		Expect(empty).To(BeEmpty())

		one := []track{{id: "1", artist: "A", album: "X"}}
		shuffle.Slice(one, keys)
		Expect(ids(one)).To(Equal([]string{"1"}))
	})

	It("permutes two items without dropping or duplicating", func() {
		tracks := []track{
			{id: "1", artist: "A", album: "X"},
			{id: "2", artist: "B", album: "Y"},
		}
		original := ids(tracks)
		shuffle.Slice(tracks, keys)
		Expect(ids(tracks)).To(ConsistOf(original))
	})

	It("is a full permutation of a larger list", func() {
		tracks := make([]track, 20)
		for i := range tracks {
			tracks[i] = track{
				id:     strconv.Itoa(i),
				artist: "Artist" + strconv.Itoa(i%5),
				album:  "Album" + strconv.Itoa(i%7),
			}
		}
		original := ids(tracks)
		shuffle.Slice(tracks, keys)
		Expect(ids(tracks)).To(ConsistOf(original))
	})

	It("still shuffles a single-artist playlist", func() {
		tracks := []track{
			{id: "1", artist: "A", album: "X"},
			{id: "2", artist: "A", album: "X"},
			{id: "3", artist: "A", album: "X"},
			{id: "4", artist: "A", album: "X"},
		}
		shuffle.Slice(tracks, keys)
		Expect(ids(tracks)).To(ConsistOf("1", "2", "3", "4"))
		Expect(adjacentSame(tracks, func(t track) string { return t.artist })).To(Equal(3))
	})

	It("produces no adjacent same-artist pairs when artists are balanced", func() {
		original := make([]track, 0, 24)
		for i := range 24 {
			original = append(original, track{
				id:     strconv.Itoa(i),
				artist: "Artist" + strconv.Itoa(i%6),
				album:  "Album" + strconv.Itoa(i),
			})
		}
		for range 25 {
			tracks := clone(original)
			shuffle.Slice(tracks, keys)
			Expect(ids(tracks)).To(ConsistOf(ids(original)))
			Expect(adjacentSame(tracks, func(t track) string { return t.artist })).To(Equal(0))
		}
	})

	It("meets the minimum adjacent-artist count when one artist dominates", func() {
		// 11 of A, 3 of B, 2 of C → n=16, maxFreq=11, min adjacent = 2*11-16-1 = 5
		original := make([]track, 0, 16)
		for i := range 11 {
			original = append(original, track{id: "A" + strconv.Itoa(i), artist: "A", album: "X"})
		}
		for i := range 3 {
			original = append(original, track{id: "B" + strconv.Itoa(i), artist: "B", album: "Y"})
		}
		for i := range 2 {
			original = append(original, track{id: "C" + strconv.Itoa(i), artist: "C", album: "Z"})
		}
		for range 20 {
			tracks := clone(original)
			shuffle.Slice(tracks, keys)
			Expect(ids(tracks)).To(ConsistOf(ids(original)))
			Expect(adjacentSame(tracks, func(t track) string { return t.artist })).To(Equal(minAdjacent(16, 11)))
		}
	})

	It("spaces albums when every track is the same artist", func() {
		original := make([]track, 0, 12)
		for i := range 12 {
			original = append(original, track{
				id:     strconv.Itoa(i),
				artist: "A",
				album:  "Album" + strconv.Itoa(i%3),
			})
		}
		for range 20 {
			tracks := clone(original)
			shuffle.Slice(tracks, keys)
			Expect(ids(tracks)).To(ConsistOf(ids(original)))
			Expect(adjacentSame(tracks, func(t track) string { return t.album })).To(Equal(0))
		}
	})
})

File: ui/src/utils/shuffle.test.js
import { describe, it, expect } from 'vitest'
import { balancedShuffle, songKeys } from './shuffle'

const ids = (tracks) => tracks.map((t) => t.id)

const adjacentSame = (tracks, field) => {
  let n = 0
  for (let i = 1; i < tracks.length; i++) {
    if (field(tracks[i]) && field(tracks[i]) === field(tracks[i - 1])) {
      n++
    }
  }
  return n
}

const minAdjacent = (n, maxFreq) => {
  if (n === 0) {
    return 0
  }
  return Math.max(0, 2 * maxFreq - n - 1)
}

describe('songKeys', () => {
  it('prefers artist, then singer, then albumArtist', () => {
    expect(
      songKeys({ artist: 'A', singer: 'S', albumArtist: 'AA', album: 'X' }),
    ).toEqual({
      artist: 'A',
      album: 'X',
    })
    expect(songKeys({ singer: 'S', albumArtist: 'AA' })).toEqual({
      artist: 'S',
      album: '',
    })
    expect(songKeys({ albumArtist: 'AA', album: 'Y' })).toEqual({
      artist: 'AA',
      album: 'Y',
    })
    expect(songKeys({})).toEqual({ artist: '', album: '' })
  })
})

describe('balancedShuffle', () => {
  it('leaves empty and single-item lists unchanged', () => {
    expect(balancedShuffle([])).toEqual([])
    expect(balancedShuffle([{ id: '1', artist: 'A', album: 'X' }])).toEqual([
      { id: '1', artist: 'A', album: 'X' },
    ])
  })

  it('permutes two items without dropping or duplicating', () => {
    const tracks = [
      { id: '1', artist: 'A', album: 'X' },
      { id: '2', artist: 'B', album: 'Y' },
    ]
    expect(ids(balancedShuffle(tracks, songKeys))).toEqual(
      expect.arrayContaining(['1', '2']),
    )
    expect(balancedShuffle(tracks, songKeys)).toHaveLength(2)
  })

  it('does not mutate the input array', () => {
    const tracks = [
      { id: '1', artist: 'A', album: 'X' },
      { id: '2', artist: 'B', album: 'Y' },
      { id: '3', artist: 'C', album: 'Z' },
    ]
    const snapshot = tracks.slice()
    balancedShuffle(tracks, songKeys)
    expect(tracks).toEqual(snapshot)
  })

  it('is a full permutation of a larger list', () => {
    const tracks = Array.from({ length: 20 }, (_, i) => ({
      id: String(i),
      artist: `Artist${i % 5}`,
      album: `Album${i % 7}`,
    }))
    const shuffled = balancedShuffle(tracks, songKeys)
    expect(ids(shuffled).sort()).toEqual(ids(tracks).sort())
  })

  it('still shuffles a single-artist playlist', () => {
    const tracks = [
      { id: '1', artist: 'A', album: 'X' },
      { id: '2', artist: 'A', album: 'X' },
      { id: '3', artist: 'A', album: 'X' },
      { id: '4', artist: 'A', album: 'X' },
    ]
    const shuffled = balancedShuffle(tracks, songKeys)
    expect(ids(shuffled).sort()).toEqual(['1', '2', '3', '4'])
    expect(adjacentSame(shuffled, (t) => t.artist)).toBe(3)
  })

  it('produces no adjacent same-artist pairs when artists are balanced', () => {
    const original = Array.from({ length: 24 }, (_, i) => ({
      id: String(i),
      artist: `Artist${i % 6}`,
      album: `Album${i}`,
    }))
    for (let n = 0; n < 25; n++) {
      const shuffled = balancedShuffle(original, songKeys)
      expect(ids(shuffled).sort()).toEqual(ids(original).sort())
      expect(adjacentSame(shuffled, (t) => t.artist)).toBe(0)
    }
  })

  it('keeps adjacent-artist count near the minimum when one artist dominates', () => {
    const original = [
      ...Array.from({ length: 11 }, (_, i) => ({
        id: `A${i}`,
        artist: 'A',
        album: 'X',
      })),
      ...Array.from({ length: 3 }, (_, i) => ({
        id: `B${i}`,
        artist: 'B',
        album: 'Y',
      })),
      ...Array.from({ length: 2 }, (_, i) => ({
        id: `C${i}`,
        artist: 'C',
        album: 'Z',
      })),
    ]
    const floor = minAdjacent(16, 11)
    for (let n = 0; n < 20; n++) {
      const shuffled = balancedShuffle(original, songKeys)
      expect(ids(shuffled).sort()).toEqual(ids(original).sort())
      // Adjacent repair is heuristic; stay near the theoretical floor.
      expect(adjacentSame(shuffled, (t) => t.artist)).toBeLessThanOrEqual(floor + 4)
    }
  })

  it('does not cycle only the top artists in a skewed library', () => {
    const original = [
      ...Array.from({ length: 49 }, (_, i) => ({
        id: `U${i}`,
        artist: 'Unknown Artist',
        album: 'U',
      })),
      ...Array.from({ length: 24 }, (_, i) => ({
        id: `AM${i}`,
        artist: 'Arctic Monkeys',
        album: 'AM',
      })),
      ...Array.from({ length: 19 }, (_, i) => ({
        id: `M${i}`,
        artist: 'Muse',
        album: 'M',
      })),
      ...Array.from({ length: 40 }, (_, i) => ({
        id: `O${i}`,
        artist: `Other${i}`,
        album: `O${i}`,
      })),
    ]
    let otherHits = 0
    for (let n = 0; n < 30; n++) {
      const shuffled = balancedShuffle(original, songKeys)
      const first20 = shuffled.slice(0, 20).map((t) => t.artist)
      if (first20.some((a) => a.startsWith('Other'))) {
        otherHits++
      }
    }
    // Greedy largest-pile almost never put Others in the first 20; random does.
    expect(otherHits).toBeGreaterThan(20)
  })

  it('reduces adjacent same-album pairs when every track is the same artist', () => {
    const original = Array.from({ length: 12 }, (_, i) => ({
      id: String(i),
      artist: 'A',
      album: `Album${i % 3}`,
    }))
    for (let n = 0; n < 20; n++) {
      const shuffled = balancedShuffle(original, songKeys)
      expect(ids(shuffled).sort()).toEqual(ids(original).sort())
      expect(adjacentSame(shuffled, (t) => t.album)).toBeLessThanOrEqual(2)
    }
  })
})
